src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,893 @@
+    bool some_local_not_tracked = false;
+    // true iff root_idents contains only local that are not tracked
+    bool all_local_not_tracked = true;
+    // true iff root_idents contains only local that are tracked
+    bool all_local_tracked = true;
+    for(const auto &root_ident : root_idents)
+    {
+      bool loc = local.find(root_ident) != local.end();
+      bool tra = tracked.find(root_ident) != tracked.end();
+      bool local_tracked = loc && tra;
+      bool local_not_tracked = loc && !tra;
+      some_non_local |= !loc;
+      some_local_not_tracked |= local_not_tracked;
+      all_local_not_tracked &= local_not_tracked;
+      all_local_tracked &= local_tracked;
+    }
+
+    // some based identifier is not local, must check
+    if(some_non_local)
+    {
+      // if we also have a local that is not tracked, we know the check will
+      // fail with the current algorithm, error out.
+      if(some_local_not_tracked)
+      {
+        throw analysis_exceptiont(
+          "LHS expression `" + from_expr(lhs) +
+          "` in assignment mentions both explicitly and implicitly tracked "
+          "memory locations. DFCC does not yet handle that case, please "
+          "reformulate the assignment into separate assignments to either "
+          "memory locations.");
+      }
+      return true;
+    }
+    else
+    {
+      // all base idents are local
+      // if they are all not tracked, we *have* to skip the check
+      // if they are all tracked, we *can* skip the check
+      // in both cases it is sound to do so.
+      if(all_local_not_tracked || all_local_tracked)
+      {
+        return false;
+      }
+      else
+      {
+        // we have a combination of tracked and not-tracked locals, we know
+        // the check will fail with the current algorithm, error out.
+        throw analysis_exceptiont(
+          "LHS expression `" + from_expr(lhs) +
+          "` in assignment mentions both explicitly and implicitly tracked "
+          "memory locations. DFCC does not yet handle that case, please "
+          "reformulate the assignment into separate assignments to either "
+          "memory locations.");
+      }
+    }
+  }
+}
+
+bool dfcc_cfg_infot::must_check_lhs(goto_programt::const_targett target) const
+{
+  PRECONDITION(target->is_assign() || target->is_function_call());
+
+  const exprt &lhs =
+    target->is_assign() ? target->assign_lhs() : target->call_lhs();
+
+  if(lhs.is_nil())
+    return false;
+
+  auto loop_id_opt = dfcc_get_loop_id(target);
+  PRECONDITION(
+    loop_id_opt.has_value() &&
+    is_valid_loop_or_top_level_id(loop_id_opt.value()));
+  auto loop_id = loop_id_opt.value();
+
+  if(is_top_level_id(loop_id))
+  {
+    return __must_check_lhs(lhs, top_level_local, top_level_tracked, log, ns);
+  }
+  else
+  {
+    auto &loop_info = loop_info_map.at(loop_id);
+    return __must_check_lhs(lhs, loop_info.local, loop_info.tracked, log, ns);
+  }
```

Comment:
See above: you really want to hide this kind of branching behind an interface.

Reply:
adding three new methods to hide the branching